### PR TITLE
Make Access syntax fail to compile for unsupported values

### LINKED ISSUE #4070: The access syntax should probably detect literal atoms and, maybe, whitespace
Here is an example of a case where a bug can be introduced by accident:

```elixir
def handle_call(:foo, _from, state) do
  stuff = do_stuff()
  {:reply [{:this, :good, stuff}], state}
end
```

Note the missing comma after `:reply`.

What I think can be improved:

1. Check if the argument preceding the `[]` is a literal atom and error if it's not `nil`. Should warn or error as well in the case of `nil` as that is a no-op pretty much.
2. We could warn if there is a space between the argument and the `[]`.

### PR DESCRIPTION
**Pull request by lexmag**
Closes #4070.

### CODE REVIEW

File: lib/elixir/lib/access.ex [hunk 1]
```diff
@@ -53,7 +53,11 @@ defmodule Access do
     :maps.find(key, map)
   end
 
-  def fetch(list, key) when is_list(list) and is_atom(key) do
+  def fetch(list, key) when is_list(list) do
+    unless is_atom(key) do
+      raise ArgumentError,
+        "the Access for keywords expect the key to be an atom, got: " <> inspect(key)
+    end
```
Comment (ericmj) [inline lib/elixir/lib/access.ex]: Should be cleaner to move this to its own clause.

File: lib/elixir/src/elixir_parser.yrl [hunk 2]
```diff
@@ -701,8 +701,16 @@ build_fn(Op, _Stab) ->
...
 build_access(Expr, {List, Location}) ->
   Meta = meta_from_location(Location),
+  ok = assert_access_available(Meta, Expr),
   {{'.', Meta, ['Elixir.Access', get]}, Meta, [Expr, List]}.
 
+assert_access_available(_Meta, nil) -> ok;
+assert_access_available(Meta, Expr)
+when is_binary(Expr); is_number(Expr); is_atom(Expr);
+     element(1, Expr) == '{}'; element(1, Expr) == '<<>>' ->
+  throw(Meta, "the Access is not available for the literal before: ", "'['");
```
Comment (josevalim) [inline lib/elixir/src/elixir_parser.yrl]: The token should not be `"'['"`. I think we need to raise a regular error and not a parser error. Another option is to not change here and change in elixir_rewrite to raise on `Access.get`.

File: lib/elixir/lib/access.ex (hunk 1 above)
Comment (lexmag) [inline lib/elixir/lib/access.ex]: I've made it at first place but decided to keep `list`-related logic grouped, but have no strong feelings one way or the other. :smile:

File: lib/elixir/src/elixir_parser.yrl (hunk 2 above)
Comment (lexmag) [inline lib/elixir/src/elixir_parser.yrl]: Right, `elixir_rewrite` seems a better place, and then we can warn about `nil` no-op case.

Comment (josevalim) [inline lib/elixir/src/elixir_parser.yrl]: Ok, let's move it there. But don't warn for nil. `nil[:foo]` can be generated at compile time after all.

Comment (lexmag) [inline lib/elixir/src/elixir_parser.yrl]: Of course `nil` inlining. :smile:

File: lib/elixir/src/elixir_rewrite.erl [hunk 3]
```diff
@@ -173,9 +176,16 @@ inline(_, _, _) -> false.
...
 %% Complex rewrite rules
 
-rewrite(?string_chars, _DotMeta, 'to_string', _Meta, [String]) when is_binary(String) ->
+rewrite(?access, _DotMeta, 'get', _Meta, [nil, Arg], _Env)
+when ?is_literal(Arg) orelse (is_atom(element(1, Arg)) andalso element(3, Arg) == nil) ->
+  nil;
+rewrite(?access, _DotMeta, 'get', Meta, [Arg, _], Env)
+when ?is_literal(Arg) orelse element(1, Arg) == '{}' orelse element(1, Arg) == '<<>>' ->
+  elixir_errors:compile_error(Meta, ?m(Env, file),
+    "the Access is not available for the value: ~ts", ['Elixir.Macro':to_string(Arg)]);
```
Comment (josevalim) [inline lib/elixir/src/elixir_rewrite.erl]: "the Access syntax"

File: lib/elixir/src/elixir_rewrite.erl [hunk 4]
```diff
@@ -173,9 +176,16 @@ inline(_, _, _) -> false.
 
 %% Complex rewrite rules
 
-rewrite(?string_chars, _DotMeta, 'to_string', _Meta, [String]) when is_binary(String) ->
+rewrite(?access, _DotMeta, 'get', _Meta, [nil, Arg], _Env)
+when ?is_literal(Arg) orelse (is_atom(element(1, Arg)) andalso element(3, Arg) == nil) ->
```
Comment (josevalim) [inline lib/elixir/src/elixir_rewrite.erl]: indent with 4 spaces

File: lib/elixir/src/elixir_rewrite.erl [hunk 5]
```diff
@@ -173,9 +176,16 @@ inline(_, _, _) -> false.
...
+  nil;
+rewrite(?access, _DotMeta, 'get', Meta, [Arg, _], Env)
+when ?is_literal(Arg) orelse element(1, Arg) == '{}' orelse element(1, Arg) == '<<>>' ->
```
Comment (josevalim) [inline lib/elixir/src/elixir_rewrite.erl]: indent with for spaces

**Comment by josevalim** [comment]
I've added just tiny comments, :shipit:

File: lib/elixir/src/elixir_rewrite.erl (hunk 3 above)
Comment (lexmag) [inline lib/elixir/src/elixir_rewrite.erl]: But this place handles not only Access syntax `[]` but also actual Access calls.

Comment (josevalim) [inline lib/elixir/src/elixir_rewrite.erl]: It will be alright

Comment (josevalim) [inline lib/elixir/src/elixir_rewrite.erl]: Or if you want, be explicit: "the Access syntax and calls to Access.get/2 are not available for..."

Comment (lexmag) [inline lib/elixir/src/elixir_rewrite.erl]: It could be strange to say "Access syntax" for `Access.get(1, :foo)` call.

Comment (josevalim) [inline lib/elixir/src/elixir_rewrite.erl]: "the Access syntax and calls to Access.get/2 are not available for the value: 1"

**Pull request merged**